### PR TITLE
[BUZZOK-3910] Added GPU enabled python311_genai

### PR DESCRIPTION
**Pull request by mkavf**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
This PR adds environment that is same as existing python311_genai but have version of torch library compiled for CUDA instead of CPU.

## Rationale
Creating embeddings with custom models that uses only CPU is slow. We want user to be able to leverage GPU's that they have to speed up this process.

## Changes
In `Dockerfile` changed from `https://download.pytorch.org/whl/cpu` to `https://download.pytorch.org/whl/cu118`

### CODE REVIEW

File: public_dropin_gpu_environments/python311_genai/Dockerfile
Comment (mkavf) [inline public_dropin_gpu_environments/python311_genai/Dockerfile:17]: Changed from `https://download.pytorch.org/whl/cpu` in regular `python311_genai`  to https://download.pytorch.org/whl/cu118 in GPU enabled one

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (mkavf) [inline public_dropin_gpu_environments/python311_genai/README.md:15]: Not sure about this. Should we try to use `faiss-gpu` my understanding that there restriction about what cuda version it supports.

**Review by elatt — COMMENTED** [review]
FYI, putting these files here will make them owned by @datarobot/tracking-agent team so can you please update CODEOWNERS for this specific dir (probably fine to keep it in the high level directory because it makes sense from a user perspective).

File: public_dropin_gpu_environments/python311_genai/requirements.txt
Comment (mkavf) [inline public_dropin_gpu_environments/python311_genai/requirements.txt:10]: Not sure about this. Should we try to use faiss-gpu my understanding that there restriction about what cuda version it supports.

Comment (yakov-g) [reply public_dropin_gpu_environments/python311_genai/requirements.txt:10]: Now it's my turn :)
 I see you put `custom-models` as owner. I disagree :) Please make it buzok

File: public_dropin_gpu_environments/python311_genai/dr_requirements.txt
Comment (yakov-g) [inline public_dropin_gpu_environments/python311_genai/dr_requirements.txt:2]: Don't you want to bump DRUM to the latest?

File: public_dropin_gpu_environments/python311_genai/env_info.json
Comment (yakov-g) [inline public_dropin_gpu_environments/python311_genai/env_info.json:3]: Please change the name and description, otherwise it makes it the same as existing one.

File: public_dropin_gpu_environments/python311_genai/dr_requirements.txt
Comment (mkavf) [reply public_dropin_gpu_environments/python311_genai/dr_requirements.txt:2]: I have no idea. This is copy of `python311_genai` with min changes to support GPU. `python311_genai` currently marked as owned by `custom-models`.

File: public_dropin_gpu_environments/python311_genai/env_info.json
Comment (yakov-g) [inline public_dropin_gpu_environments/python311_genai/env_info.json:2]: change ID

File: public_dropin_gpu_environments/python311_genai/env_info.json
Comment (mkavf) [reply public_dropin_gpu_environments/python311_genai/env_info.json:3]: Is there best practise how to mark that this env is for GPU in the name ?

Comment (yakov-g) [reply public_dropin_gpu_environments/python311_genai/env_info.json:3]: Probably yes. Just to make it clear

Comment (yakov-g) [reply public_dropin_gpu_environments/python311_genai/env_info.json]: Make it clear in the name and details in description.

File: public_dropin_gpu_environments/python311_genai/env_info.json
Comment (yakov-g) [reply public_dropin_gpu_environments/python311_genai/env_info.json]: and change environment version id

Comment (mkavf) [reply public_dropin_gpu_environments/python311_genai/env_info.json]: to what version ?

Comment (mkavf) [reply public_dropin_gpu_environments/python311_genai/env_info.json]: any ?

Comment (yakov-g) [reply public_dropin_gpu_environments/python311_genai/env_info.json]: python3 -c "import bson; print(bson.ObjectId())"

Comment (yakov-g) [reply public_dropin_gpu_environments/python311_genai/env_info.json]: There are fields `id` and `environmentVersionId`.
Every environment has an ID. It always stays the same, doesn't change with the environment bump.
When we do changes, we need to bump environment version ID, to make it a different version.

Just generate any ObjectId, using bson.

File: public_dropin_gpu_environments/python311_genai/requirements.txt
Comment (yakov-g) [reply public_dropin_gpu_environments/python311_genai/requirements.txt:10]: Custom Models can not be an owner, we know nothing about this env.
`python311_genai` will likelyl be owned by @baekdahl

**Comment by yakov-g** [comment]
Also I'd think of some tests for the newly added envs.
@elatt for GPU envs as well.
Unless you have some other tests not reporting to this repo.

File: public_dropin_gpu_environments/python311_genai/env_info.json
Comment (mkavf) [inline public_dropin_gpu_environments/python311_genai/env_info.json:4]: @jendavies How best rewrite following:  "This environment is expected to be run on machines with GPU"

File: DRCODEOWNERS
Comment (elatt) [inline DRCODEOWNERS:45]: @yakov-g with only having a single team reviewing, can we change this repo to only require a single reviewer?

**Comment by elatt** [comment]
> Also I'd think of some tests for the newly added envs. @elatt for GPU envs as well. Unless you have some other tests not reporting to this repo.

We have some tests: https://github.com/datarobot/datarobot-user-models/blob/master/tests/functional/test_inference_per_framework.py#L1166

File: public_dropin_gpu_environments/python311_genai/requirements.txt
Comment (baekdahl) [reply public_dropin_gpu_environments/python311_genai/requirements.txt:10]: Yes, I can own this.
@mkavf What is the point of this environment? To me, it looks identical to the non-gpu python311_genai.

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/README.md:7]:
```suggestion
This environment is built for Python 3 and has support for the following libraries.
```

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/README.md:8]:
```suggestion
For specific version information and the complete list of included packages, see the [requirements](requirements.txt).
```

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/README.md:29]:
```suggestion
2. Using either the API or from the UI, create a new custom environment with the tarball created
in step 1.
```

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/README.md:1]:
```suggestion
# Python 3 GenAI drop-in template environment
```

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/README.md:33]:
```suggestion
To use this environment, your custom model archive will typically contain a `custom.py` file containing the necessary hooks, as well as other files needed for your workflow. You can implement the hook functions such as `load_model` and `score_unstructured`, as documented [here](../../custom_model_runner/README.md).
```

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/README.md:35]:
```suggestion
Implement your Python code under the related custom hook functions to build your GenAI workflows. Within your `custom.py` code, import the necessary dependencies found in this environment.
```

File: public_dropin_gpu_environments/python311_genai/README.md
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/README.md:37]:
```suggestion
If you need additional dependencies, add those packages in the `requirements.txt` file that you include within your custom model archive. DataRobot will make them available to your custom Python code after you build the environment.
```

File: public_dropin_gpu_environments/python311_genai/env_info.json
Comment (jendavies) [inline public_dropin_gpu_environments/python311_genai/env_info.json:4]:
```suggestion
  "description": "This template environment can be used to create GenAI-powered custom models and includes common dependencies for workflows using OpenAI, Langchain, vector DBs, or transformers in PyTorch. This environment is created to run on machines with GPU support. Similar to other drop-in environments, you can either include a `.pth` artifact or any other code needed to deserialize your model, and optionally, a `custom.py` file.",
```

**Review by jendavies — APPROVED** [review]
Looks good!

File: public_dropin_gpu_environments/python311_genai/requirements.txt
Comment (mkavf) [reply public_dropin_gpu_environments/python311_genai/requirements.txt:10]: @baekdahl 

It is almost identical. In Docketfile I changed `https://download.pytorch.org/whl/cpu` in regular python311_genai to `https://download.pytorch.org/whl/cu118`. This will allow us to utilize  GPU's.

Also I considering changing `faiss-cpu` to `faiss-gpu` but now sure what version of CUDA we support or should support

Comment (mkavf) [reply public_dropin_gpu_environments/python311_genai/requirements.txt:10]: Seem there no corresponding version of  `faiss-gpu` for our version, so i will leave `faiss-cpu` as is.

File: DRCODEOWNERS
Comment (yakov-g) [reply DRCODEOWNERS:45]: We historically required two because it is public repo. Just to have another pair of eyes.
Generally we could remove, but I'm not sure.

Comment (yakov-g) [reply DRCODEOWNERS:45]: @baekdahl @mkavf Maybe you could update ownership for it accordingly. And for the original genai env as well.

**Review by yakov-g — APPROVED** [review]
Please update ownership , to have @baekdahl 's team which is @datarobot/codegen
I believe

**Comment by yakov-g** [comment]
@mkavf ping me when you'll be ready to install it. I'll tell you how to use installer if you don''t know yet.

**Review by baekdahl — APPROVED** [review]

**Comment by mkavf** [comment]
> @mkavf ping me when you'll be ready to install it. I'll tell you how to use installer if you don''t know yet.

@yakov-g 
No, don't know how to use installer. Why do I need it ? To add this environment to already running application ?

**Comment by mkavf** [comment]
@elatt Is there anything else i need to do to before merge.

**Pull request merged**
